src/arduino.cc/arduino-builder/main.go

```diff
@@ -170,85 +172,77 @@ func main() {
 	}
 
 	context := make(map[string]interface{})
+	ctx := &types.Context{}
 
-	buildOptions := make(map[string]string)
 	if *buildOptionsFileFlag != "" {
+		buildOptions := make(props.PropertiesMap)
 		if _, err := os.Stat(*buildOptionsFileFlag); err == nil {
 			data, err := ioutil.ReadFile(*buildOptionsFileFlag)
 			if err != nil {
 				printCompleteError(err)
-				defer os.Exit(1)
-				return
 			}
 			err = json.Unmarshal(data, &buildOptions)
 			if err != nil {
 				printCompleteError(err)
-				defer os.Exit(1)
-				return
 			}
 		}
+		ctx.InjectBuildOptions(buildOptions)
 	}
 
-	var err error
-	printStackTrace := false
-	err, printStackTrace = setContextSliceKeyOrLoadItFromOptions(context, hardwareFoldersFlag, buildOptions, constants.CTX_HARDWARE_FOLDERS, FLAG_HARDWARE, true)
-	if err != nil {
-		printError(err, printStackTrace)
-		defer os.Exit(1)
-		return
+	// FLAG_HARDWARE
+	if hardwareFolders, err := toSliceOfUnquoted(hardwareFoldersFlag); err != nil {
+		printCompleteError(err)
+	} else if len(hardwareFolders) > 0 {
+		ctx.HardwareFolders = hardwareFolders
 	}
-
-	err, printStackTrace = setContextSliceKeyOrLoadItFromOptions(context, toolsFoldersFlag, buildOptions, constants.CTX_TOOLS_FOLDERS, FLAG_TOOLS, true)
-	if err != nil {
-		printError(err, printStackTrace)
-		defer os.Exit(1)
-		return
+	if len(ctx.HardwareFolders) == 0 {
+		printErrorMessageAndFlagUsage(errors.New("Parameter '" + FLAG_HARDWARE + "' is mandatory"))
 	}
 
-	err, printStackTrace = setContextSliceKeyOrLoadItFromOptions(context, librariesFoldersFlag, buildOptions, constants.CTX_OTHER_LIBRARIES_FOLDERS, FLAG_LIBRARIES, false)
-	if err != nil {
-		printError(err, printStackTrace)
-		defer os.Exit(1)
-		return
+	// FLAG_TOOLS
+	if toolsFolders, err := toSliceOfUnquoted(toolsFoldersFlag); err != nil {
+		printCompleteError(err)
+	} else if len(toolsFolders) > 0 {
+		ctx.ToolsFolders = toolsFolders
 	}
-
-	err, printStackTrace = setContextSliceKeyOrLoadItFromOptions(context, librariesBuiltInFoldersFlag, buildOptions, constants.CTX_BUILT_IN_LIBRARIES_FOLDERS, FLAG_BUILT_IN_LIBRARIES, false)
-	if err != nil {
-		printError(err, printStackTrace)
-		defer os.Exit(1)
-		return
+	if len(ctx.ToolsFolders) == 0 {
+		printErrorMessageAndFlagUsage(errors.New("Parameter '" + FLAG_TOOLS + "' is mandatory"))
 	}
 
-	err, printStackTrace = setContextSliceKeyOrLoadItFromOptions(context, customBuildPropertiesFlag, buildOptions, constants.CTX_CUSTOM_BUILD_PROPERTIES, FLAG_PREFS, false)
-	if err != nil {
-		printError(err, printStackTrace)
-		defer os.Exit(1)
-		return
+	// FLAG_LIBRARIES
+	if librariesFolders, err := toSliceOfUnquoted(librariesFoldersFlag); err != nil {
+		printCompleteError(err)
+	} else if len(librariesFolders) > 0 {
+		ctx.LibrariesFolders = librariesFolders
```

Comment:
There is a bug here, this should be `ctx.OtherLibrariesFolder` instead. `ctx.LibrariesFolders` is used by `LoadLibraries` to store the final complete and sorted list of library folders.